File: time_offset/codes/Gardner_Time_sync.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar  8 15:22:28 2019

@author: theresh
"""
import numpy as np
from matplotlib import pyplot as plt
from scipy import special

N=100000
data=np.random.randint(0,2,N)
c=2*data-1
Tsym=100
#sym_samp=np.arange(0,Tsym)
Eb_N0_dB=np.linspace(0,8,9)
BER_sim=np.zeros(len(Eb_N0_dB))
pulse=np.ones(Tsym)
inter_data=np.zeros(N*Tsym)
for i in range(N):
    inter_data[i*Tsym:(i+1)*Tsym]=c[i]
#plt.plot(inter_data)
#plt.show()

#toff=np.zeros(len(Eb_N0_dB))
#poff=np.zeros(len(Eb_N0_dB))
for i in range(len(Eb_N0_dB)):
    N0 = 1/(np.exp(Eb_N0_dB[i]*np.log(10)/10.0))
    noise=np.random.normal(0,np.sqrt(N0/2.0),len(inter_data))
    rx=inter_data+noise
    tau=0
    delta=int(Tsym/2)
    center=60   # A r b i t a r y  r e c e i v e d  c e n t e r index
    #toff[i]=center%delta
    a=np.zeros(N)    
    avgsamples=6
    stepsize=1
    rit=-1
    GA=np.zeros(avgsamples)
    for k in range((delta),len(rx)-(delta),Tsym):
        #print k,
        rit=rit+1
        #b.append(rit)
        midsample=rx[center-1]
        latesample=rx[center+delta-1]
        earlysample=rx[center-delta-1]
        a[rit]=earlysample
        sub=latesample-earlysample
        GA[np.mod(rit,avgsamples)]=sub*midsample
        if (np.mean(GA)>0):
            tau=-stepsize
        else:
            tau=stepsize
        center=center+Tsym+tau
        if center>=len(rx)-delta:
            break;
    #print center
    #dum=np.fmod(center,delta)
    #poff[i]=dum
    xc=1*(a>0)
    error=(xc!=data).sum()
    BER_sim[i]=(1.0*error)/N 
    
theoryBer = 0.5*special.erfc(np.sqrt(10.0**(Eb_N0_dB/10.0)))    
plt.semilogy(Eb_N0_dB,theoryBer)    
plt.semilogy(Eb_N0_dB,BER_sim) 
plt.title("Timing Sync using Gardner TED")
plt.legend(['Theory','Simulated'],loc='best')
plt.xlabel('Eb/No (dB)')
plt.ylabel('$P_e$')
plt.grid(True)
plt.show()

File: synctech/codes/Time_Synchronization/Matlab/Gardner_time_sync.m
clc;
clear all;
close all;
%=========================================================================================
% This program implemented based on the ""A BPSK/QPSK Timing-Error Detector for Sampled
% Receivers by FLOYD M.GARDNER,IEEE TRANSACTIONS ON COMMUNICATIONS, VOL.COM-34,NO. 5,
% MAY 1986"". This is a simple algorithm for detection of timing error of a synchronous,
% bandlimited, BPSK or QPSK datastrem. The algorithm requires two samples for symbol. And
% uses one of the two samples for the Symbol decision. It's purely a non data aided Timing
% Error detection.
%=========================================================================================



%=========================================================================================
% Inputs : 1) Simulation Length (N) (Total number of symbols)
%	   2) Samples per Symbol (M)
%          3) Eb_N0_dB Range
%          4) Length_Prev_Symbols (For averange purpose)
%
% Output : 1) BER_sim

% Plot : BER Curve for theorical simulation vs Algorithm             
%==========================================================================================

%==========================================================================================
%                                      Transmitter
%==========================================================================================

N=100000 ;   % Simulation length
data=rand(1,N)>0.5; % Generation of random bits
c=2*data-1;  % BPSK mapping
M=100;     % Number of samples per symbol.
p=ones(1,M);            % shape Pulse shape
Mapped_Data=zeros(1,N*M); % Total samples
Mapped_Data(1:M:end)=c ; %Interpolated BPSK data using the pulse shaping filter(NRZ)
Mapped_Data=conv(Mapped_Data,p);

%%%%==========================================================================================
%%%% Noise addition
Eb_N0_dB=1:1:10; % Signal to Noise Ratio
len_eb_n0=length(Eb_N0_dB);        
time_offsets=([3,50,80]);  % List of Timing Offsets
len_time_offsets=length(time_offsets);
BER_sim=zeros(len_time_offsets,len_eb_n0) ;  % Bit Error Rate Values
 for off=1:len_time_offsets
     for i=1:len_eb_n0
         %========================================================
         % Noise characteristics
         N0 = 10^(Eb_N0_dB(i)/10);
         noise=sqrt(1/(2*N0))*randn(1,length(Mapped_Data));
         %=========================================================
         rx=Mapped_Data+noise;   % Noisy received samples
         delta=M/2;
         timing_offset=time_offsets(off);  % Timing offset addition 
         center=delta + timing_offset;  % New center
         %===========================================================
         %                 Gardner TED algorithm
         %===========================================================
         Decision_Samples=zeros(1,N); %Samples storing for making Decision. (DS)
         Length_Prev_Symbols=6;     %User Specified Le: Used for the Averaging Purpose
         Prev_Samples=zeros(1,Length_Prev_Symbols);
         increment=1;
         index_DS=0;   % Index of Decision samples while storing the data.
         tau=0;
         for k=delta:M:length(rx)-delta
             index_DS=index_DS+1;
             m_s=rx(center);   % Middle sample
             l_s=rx(center+delta); % Late sample
             e_s=rx(center-delta); % Early sample
             Decision_Samples(index_DS)=e_s ; %As per the algorithm.
             sub=l_s-e_s;
             Prev_Samples(mod(index_DS,Length_Prev_Symbols)+1)=sub*m_s ; % Algorithm specified
             if mean(Prev_Samples)>0
                 tau=-increment;
             else
                 tau=increment;
             end
             center=center+M+tau;
             if center>=length(rx)- delta-1
                 break;
             end               
         end
         
             %===============================================================
             
                   % Decision and Bit Error Rate Calulation
     Error=0; %Set the initial value for Error         
    for k=1:N-1 %Error calculation
        if ((Decision_Samples(k)>0 && c(k)==-1)||(Decision_Samples(k)< 0 && c(k)==1))
            Error=Error+1;
        end
    end
    BER_sim(off,i)=Error/(N-1);
    end
 end
BER_th=(1/2)*erfc(sqrt(10.^(Eb_N0_dB/10))); %Calculate The theoretical BER
figure
semilogy(Eb_N0_dB,BER_th); %Plot theoretical BER
hold on
semilogy(Eb_N0_dB,BER_sim(1,:)); %Plot simulated BER
hold on
semilogy(Eb_N0_dB,BER_sim(2,:)); %Plot simulated BER
hold on
semilogy(Eb_N0_dB,BER_sim(3,:)); %Plot simulated BER
title('SNR Vs. BER for BPSK­Gardner technique');
legend('Theoritical','\tau=3','\tau=50','\tau=80','Interpreter','latex');
xlabel('$\frac{E_b}{N_0}$(dB)','Interpreter','latex')
ylabel('$P_e$','Interpreter','latex')